modules/host/negotiatedownload.go

```diff
@@ -0,0 +1,230 @@
+		}
+		if totalSize > settings.MaxDownloadBatchSize {
+			return errLargeDownloadBatch
+		}
+
+		// Verify that the correct amount of money has been moved from the
+		// renter's contract funds to the host's contract funds.
+		expectedTransfer := settings.MinimumDownloadBandwidthPrice.Mul(types.NewCurrency64(totalSize))
+		existingRevision := so.RevisionTransactionSet[len(so.RevisionTransactionSet)-1].FileContractRevisions[0]
+		// Check that NewValidProofOutputs[0] is deducted by expectedTransfer.
+		if paymentRevision.NewValidProofOutputs[0].Value.Add(expectedTransfer).Cmp(existingRevision.NewValidProofOutputs[0].Value) != 0 {
+			return errors.New("bad payments")
+		}
+		// Check that NewValidProofOutputs[1] is incremented by expectedTransfer.
+		if existingRevision.NewValidProofOutputs[1].Value.Add(expectedTransfer).Cmp(paymentRevision.NewValidProofOutputs[1].Value) != 0 {
+			return errors.New("bad payouts")
+		}
+		// Check that MissedProofOutputs[0] is decremented by expectedTransfer.
+		if paymentRevision.NewMissedProofOutputs[0].Value.Add(expectedTransfer).Cmp(existingRevision.NewMissedProofOutputs[0].Value) != 0 {
+			return errors.New("bad payments")
+		}
+		// Check that NewMissedProofOutputs[2] is incremented by expectedTransfer.
+		if existingRevision.NewMissedProofOutputs[2].Value.Add(expectedTransfer).Cmp(paymentRevision.NewMissedProofOutputs[2].Value) != 0 {
+			return errors.New("bad payouts")
+		}
+		// Check that the revision count has increased.
+		if paymentRevision.NewRevisionNumber <= existingRevision.NewRevisionNumber {
+			return errors.New("bad revision number")
+		}
+
+		// Check that no other fields have changed.
+		if paymentRevision.ParentID != existingRevision.ParentID {
+			return errors.New("unstable revision")
+		}
+		if paymentRevision.NewFileSize != existingRevision.NewFileSize {
+			return errors.New("unstable revision")
+		}
+		if paymentRevision.NewFileMerkleRoot != existingRevision.NewFileMerkleRoot {
+			return errors.New("unstable revision")
+		}
+		if paymentRevision.NewWindowStart != existingRevision.NewWindowStart {
+			return errors.New("unstable revision")
+		}
+		if paymentRevision.NewWindowEnd != existingRevision.NewWindowEnd {
+			return errors.New("unstable revision")
+		}
+		if paymentRevision.NewUnlockHash != existingRevision.NewUnlockHash {
+			return errors.New("unstable revision")
+		}
+		if paymentRevision.NewMissedProofOutputs[1].Value.Cmp(existingRevision.NewMissedProofOutputs[1].Value) != 0 {
+			return errors.New("unstable revision")
+		}
+
+		// Load the sectors and build the data payload.
+		for _, request := range requests {
+			sectorData, err := h.readSector(request.MerkleRoot)
+			if err != nil {
+				return err
+			}
+			payload = append(payload, sectorData[request.Offset:request.Offset+request.Length])
+		}
+
+		// Read the transaction signature from the renter.
+		var renterSignature types.TransactionSignature
+		err = encoding.ReadObject(conn, &renterSignature, 16e3)
```

Comment:
in revise, the renter sends the revision, host sends acceptance, then renter sends signatures. Here, the renter sends the revision and the signatures, then host sends acceptance. Can we standardize on one?